src/meep.hpp

```diff
@@ -386,24 +386,26 @@ class h5file {
 
   bool ok();
 
-  realnum *read(const char *dataname, int *rank, size_t *dims, int maxrank);
-  void write(const char *dataname, int rank, const size_t *dims, realnum *data,
+  void *read(const char *dataname, int *rank, size_t *dims, int maxrank,
+             bool single_precision = true);
+  void write(const char *dataname, int rank, const size_t *dims, void *data,
              bool single_precision = true);
-
   char *read(const char *dataname);
   void write(const char *dataname, const char *data);
 
   void create_data(const char *dataname, int rank, const size_t *dims, bool append_data = false,
                    bool single_precision = true);
   void extend_data(const char *dataname, int rank, const size_t *dims);
   void create_or_extend_data(const char *dataname, int rank, const size_t *dims, bool append_data,
-                             bool single_precision);
-  void write_chunk(int rank, const size_t *chunk_start, const size_t *chunk_dims, realnum *data);
+                             bool single_precision = true);
+  void write_chunk(int rank, const size_t *chunk_start, const size_t *chunk_dims, void *data,
+                   bool single_precision = true);
   void write_chunk(int rank, const size_t *chunk_start, const size_t *chunk_dims, size_t *data);
   void done_writing_chunks();
 
   void read_size(const char *dataname, int *rank, size_t *dims, int maxrank);
-  void read_chunk(int rank, const size_t *chunk_start, const size_t *chunk_dims, realnum *data);
+  void read_chunk(int rank, const size_t *chunk_start, const size_t *chunk_dims, void *data,
+                  bool single_precision = true);
```

Comment:
Better to have two functions:
```c++
void read_chunk(int rank, const size_t *chunk_start, const size_t *chunk_dims, float *data);
void read_chunk(int rank, const size_t *chunk_start, const size_t *chunk_dims, double *data);
```

Reply:
If you do it my way, the compiler picks the right method and will always get it right.  With your way, the programmer has to pass the correct value of `single_precision` and can get it wrong.

Similarly for `write_chunk`.

Reply:
Done.